File: custom-web-interface/cors-proxy-server.js
/**
 * Serveur proxy CORS pour Faraday
 * Ce serveur fait l'interface entre le frontend et l'API Faraday
 * en ajoutant les headers CORS nécessaires
 */

const http = require('http');
const https = require('https');
const url = require('url');
const path = require('path');
const fs = require('fs');

// Configuration
const PROXY_PORT = 8082;
const FARADAY_BASE_URL = 'http://localhost:5985';
const STATIC_DIR = __dirname; // Servir les fichiers statiques depuis ce répertoire

// Types de contenu
const MIME_TYPES = {
    '.html': 'text/html',
    '.js': 'text/javascript',
    '.css': 'text/css',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpg',
    '.gif': 'image/gif',
    '.ico': 'image/x-icon',
    '.svg': 'image/svg+xml'
};

function getMimeType(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    return MIME_TYPES[ext] || 'text/plain';
}

function addCorsHeaders(res) {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, PATCH, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization, X-CSRFToken, Accept');
    res.setHeader('Access-Control-Expose-Headers', 'X-CSRFToken, Authorization');
    res.setHeader('Access-Control-Allow-Credentials', 'true');
    res.setHeader('Access-Control-Max-Age', '86400');
}

function handleStaticFile(req, res, filePath) {
    // Servir les fichiers statiques
    if (!fs.existsSync(filePath)) {
        res.writeHead(404, { 'Content-Type': 'text/plain' });
        res.end('File not found');
        return;
    }

    const stat = fs.statSync(filePath);
    if (stat.isDirectory()) {
        // Essayer index.html dans le répertoire
        const indexPath = path.join(filePath, 'index.html');
        if (fs.existsSync(indexPath)) {
            filePath = indexPath;
        } else {
            res.writeHead(404, { 'Content-Type': 'text/plain' });
            res.end('Directory listing not allowed');
            return;
        }
    }

    const mimeType = getMimeType(filePath);
    const fileStream = fs.createReadStream(filePath);
    
    addCorsHeaders(res);
    res.writeHead(200, { 'Content-Type': mimeType });
    fileStream.pipe(res);
}

function proxyToFaraday(req, res) {
    console.log(`🔄 Proxying: ${req.method} ${req.url}`);

    // Traiter les requêtes OPTIONS (preflight CORS)
    if (req.method === 'OPTIONS') {
        addCorsHeaders(res);
        res.writeHead(200);
        res.end();
        return;
    }

    // Construire l'URL Faraday
    const faradayUrl = `${FARADAY_BASE_URL}${req.url}`;
    const parsedUrl = url.parse(faradayUrl);

    const options = {
        hostname: parsedUrl.hostname,
        port: parsedUrl.port || 80,
        path: parsedUrl.path,
        method: req.method,
        headers: {
            ...req.headers,
            host: parsedUrl.host
        }
    };

    // Supprimer les headers problématiques
    delete options.headers.origin;
    delete options.headers.referer;

    const proxy = http.request(options, (faradayRes) => {
        // Ajouter les headers CORS
        addCorsHeaders(res);
        
        // Copier les headers de réponse de Faraday
        Object.keys(faradayRes.headers).forEach(key => {
            res.setHeader(key, faradayRes.headers[key]);
        });

        res.writeHead(faradayRes.statusCode);
        faradayRes.pipe(res);
    });

    proxy.on('error', (err) => {
        console.error('❌ Erreur proxy:', err.message);
        addCorsHeaders(res);
        res.writeHead(502, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({
            error: 'Proxy Error',
            message: 'Impossible de contacter le serveur Faraday',
            details: err.message
        }));
    });

    // Transférer le corps de la requête
    req.pipe(proxy);
}

// Créer le serveur
const server = http.createServer((req, res) => {
    const parsedUrl = url.parse(req.url, true);
    const pathname = parsedUrl.pathname;

    console.log(`📥 ${req.method} ${pathname}`);

    // Routes API - proxy vers Faraday
    if (pathname.startsWith('/_api') || pathname.startsWith('/v3')) {
        proxyToFaraday(req, res);
        return;
    }

    // Routes statiques - servir les fichiers locaux
    let filePath = path.join(STATIC_DIR, pathname === '/' ? 'index.html' : pathname);
    
    // Sécurité - empêcher l'accès aux fichiers en dehors du répertoire
    if (!filePath.startsWith(STATIC_DIR)) {
        res.writeHead(403, { 'Content-Type': 'text/plain' });
        res.end('Access denied');
        return;
    }

    handleStaticFile(req, res, filePath);
});

// Gestion des erreurs du serveur
server.on('error', (err) => {
    if (err.code === 'EADDRINUSE') {
        console.error(`❌ Le port ${PROXY_PORT} est déjà utilisé. Essayez de fermer l'autre serveur ou utilisez un autre port.`);
    } else {
        console.error('❌ Erreur du serveur:', err);
    }
});

// Démarrer le serveur
server.listen(PROXY_PORT, () => {
    console.log('🚀 Serveur proxy CORS Faraday démarré !');
    console.log(`📡 Proxy API: http://localhost:${PROXY_PORT}/_api/v3/`);
    console.log(`🌐 Interface web: http://localhost:${PROXY_PORT}`);
    console.log(`🔗 Faraday backend: ${FARADAY_BASE_URL}`);
    console.log('');
    console.log('ℹ️  Ce serveur:');
    console.log('   - Sert votre interface web personnalisée');
    console.log('   - Redirige les appels API vers Faraday avec les headers CORS');
    console.log('   - Gère l\'authentification automatiquement');
    console.log('');
    console.log('✅ Utilisez http://localhost:8082 pour accéder à votre interface');
});

// Gestion propre de l'arrêt
process.on('SIGINT', () => {
    console.log('\n🛑 Arrêt du serveur proxy...');
    server.close(() => {
        console.log('✅ Serveur arrêté proprement');
        process.exit(0);
    });
});

process.on('SIGTERM', () => {
    console.log('\n🛑 Arrêt du serveur proxy...');
    server.close(() => {
        console.log('✅ Serveur arrêté proprement');
        process.exit(0);
    });
});

File: custom-web-interface/cors-proxy-server.py
#!/usr/bin/env python3
"""
Serveur Proxy CORS pour Faraday
Permet de contourner les restrictions CORS pour connecter l'interface web aux vraies données Faraday
"""

import http.server
import socketserver
import urllib.request
import urllib.parse
import json
import base64
from urllib.error import URLError, HTTPError

class CORSProxyHandler(http.server.BaseHTTPRequestHandler):
    
    FARADAY_BASE_URL = "http://localhost:5985"
    
    def do_OPTIONS(self):
        """Gérer les requêtes preflight CORS"""
        self.send_response(200)
        self.send_cors_headers()
        self.end_headers()
    
    def do_GET(self):
        """Proxy pour les requêtes GET"""
        self.proxy_request('GET')
    
    def do_POST(self):
        """Proxy pour les requêtes POST"""
        self.proxy_request('POST')
    
    def do_PUT(self):
        """Proxy pour les requêtes PUT"""
        self.proxy_request('PUT')
    
    def do_DELETE(self):
        """Proxy pour les requêtes DELETE"""
        self.proxy_request('DELETE')
    
    def send_cors_headers(self):
        """Envoyer les en-têtes CORS"""
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization, X-CSRFToken')
        self.send_header('Access-Control-Expose-Headers', 'X-CSRFToken')
        self.send_header('Access-Control-Allow-Credentials', 'true')
    
    def proxy_request(self, method):
        """Proxy une requête vers Faraday"""
        try:
            # Mapper les anciens endpoints vers les nouveaux
            path = self.path
            
            # Convertir _api vers _api si nécessaire
            if path.startswith('/_api/info'):
                # Essayons d'abord avec /ws (workspaces endpoint)
                path = '/ws'
            elif path.startswith('/_api/v3/workspaces'):
                path = '/ws'
            elif path.startswith('/_api/v3/ws/'):
                # Format: _api/v3/ws/workspace_name/hosts -> workspace_name/hosts
                path = path.replace('/_api/v3/ws/', '/')
            elif path.startswith('/_api/'):
                # Garder le path tel quel pour l'API Faraday
                pass
            
            # Construire l'URL Faraday finale
            faraday_url = f"{self.FARADAY_BASE_URL}{path}"
            print(f"🌐 Proxy {method} {self.path} -> {faraday_url}")
            
            # Préparer les headers pour Faraday
            headers = {}
            
            # Copier les headers importants
            if 'Authorization' in self.headers:
                headers['Authorization'] = self.headers['Authorization']
            if 'Content-Type' in self.headers:
                headers['Content-Type'] = self.headers['Content-Type']
            
            # Lire le body pour POST/PUT
            data = None
            if method in ['POST', 'PUT'] and 'Content-Length' in self.headers:
                content_length = int(self.headers['Content-Length'])
                data = self.rfile.read(content_length)
            
            # Créer la requête vers Faraday
            req = urllib.request.Request(faraday_url, data=data, headers=headers, method=method)
            
            # Exécuter la requête
            with urllib.request.urlopen(req, timeout=30) as response:
                # Envoyer la réponse au client
                self.send_response(response.getcode())
                self.send_cors_headers()
                
                # Copier les headers de réponse
                for header, value in response.headers.items():
                    if header.lower() not in ['connection', 'transfer-encoding']:
                        self.send_header(header, value)
                
                self.end_headers()
                
                # Copier le contenu
                content = response.read()
                self.wfile.write(content)
                
                print(f"✅ Proxy réussi: {response.getcode()} ({len(content)} bytes)")
        
        except HTTPError as e:
            print(f"❌ Erreur HTTP: {e.code} {e.reason}")
            self.send_response(e.code)
            self.send_cors_headers()
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            
            error_response = {
                'error': f'HTTP {e.code}: {e.reason}',
                'code': e.code
            }
            self.wfile.write(json.dumps(error_response).encode())
        
        except URLError as e:
            print(f"❌ Erreur URL: {e.reason}")
            self.send_response(500)
            self.send_cors_headers()
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            
            error_response = {
                'error': f'Connexion échouée: {e.reason}',
                'code': 500
            }
            self.wfile.write(json.dumps(error_response).encode())
        
        except Exception as e:
            print(f"❌ Erreur inattendue: {str(e)}")
            self.send_response(500)
            self.send_cors_headers()
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            
            error_response = {
                'error': f'Erreur serveur: {str(e)}',
                'code': 500
            }
            self.wfile.write(json.dumps(error_response).encode())
    
    def log_message(self, format, *args):
        """Override pour personnaliser les logs"""
        print(f"[{self.address_string()}] {format % args}")

def run_proxy_server(port=8082):
    """Démarrer le serveur proxy"""
    try:
        with socketserver.TCPServer(("", port), CORSProxyHandler) as httpd:
            print(f"🚀 Serveur Proxy CORS démarré sur http://localhost:{port}")
            print(f"📡 Proxy vers Faraday: http://localhost:5985")
            print(f"🌐 Interface web: http://localhost:8888")
            print("=" * 60)
            print("INSTRUCTIONS:")
            print("1. Ouvrez http://localhost:8888 dans votre navigateur")
            print("2. L'interface utilisera automatiquement le proxy")
            print("3. Vous aurez accès aux vraies données Faraday!")
            print("=" * 60)
            httpd.serve_forever()
    except KeyboardInterrupt:
        print("\n🛑 Serveur proxy arrêté")
    except Exception as e:
        print(f"❌ Erreur démarrage serveur: {e}")

if __name__ == "__main__":
    run_proxy_server()
